theano/gof/opt.py

```diff
@@ -1353,21 +1353,27 @@ def transform(self, node):
                     self.process_count[opt] += 1
                 if not new_repl:
                     continue
-                else:
-                    if self.profile:
-                        self.node_created[opt] += len(graph.ops(fgraph.variables, new_repl))
-                        self.applied_true[opt] += 1
-                    break  # break from the for loop over optimization.
+                if isinstance(new_repl, (tuple, list)):
+                    new_vars = new_repl
+                else:  # It must be a dict
+                    new_vars = list(new_repl.values())
+                if self.profile:
+                    self.node_created[opt] += len(graph.ops(fgraph.variables, new_vars))
+                    self.applied_true[opt] += 1
+                break  # break from the for loop over optimization.
             if not new_repl:  # No optimization applied in the last iteration
                 return repl
-            # only 1 iteration or we are at the start of the graph.
-            if not self.apply_all_opts or not new_repl[0].owner:
+            # only 1 iteration
+            if not self.apply_all_opts:
+                return new_repl
+            if not new_vars[0].owner:
+                # We are at the start of the graph.
```

Comment:
This LocalOptGroup optimization apply a list of optimization on one node. When one opt is applied, it optionaly, can try to optimize that new node imediatly. This remove sequential change to the FunctionGraph. It does mostly an equilibrium on that node.

If one optimization return as variable an input of the graph, we can't get the node that generated it. So we can't continue the "Equilibrium". This if is for that, stop when we can't continue and don't crash.